### PR TITLE
Expose Connector in the public API

### PR DESCRIPTION
**Pull request by delvedor**
As titled. I've also added an example to show how to validate the CA fingerprint.

Related: https://github.com/nodejs/undici/discussions/904 #905

### CODE REVIEW

**Review by ronag — APPROVED** [review]

**Review by ronag — CHANGES_REQUESTED** [review]
Actually isn't it better to just expose the factory method instead of the class? The class is just an implementation detail.

**Comment by delvedor** [comment]
> Actually isn't it better to just expose the factory method instead of the class? The class is just an implementation detail.

I like exposing the class, most of the time the user will just need to initialize it, but if needed it can be extended quite easily.

**Comment by ronag** [comment]
> I like exposing the class, most of the time the user will just need to initialize it, but if needed it can be extended quite easily.

But it being a class doesn't really do anything? Nothing is exposed. It's just a silly way to implement a closure.

**Review by ronag — APPROVED** [review]
I'd prefer a method but I guess a class is fine...

**Comment by delvedor** [comment]
Hold on before merging, there might be an issue with the types, checking...

File: test/ca-fingerprint.js [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+      connect (opts, cb) {
+        connector.connect(opts, (err, socket) => {
+          if (err) {
+            return cb(err)
+          }
+          if (getIssuerCertificate(socket).fingerprint256 !== 'FO:OB:AR') {
+            socket.destroy()
+            return cb(new Error('Fingerprint does not match'))
+          }
+          return cb(null, socket)
```
Comment (ronag) [inline test/ca-fingerprint.js:81]:
```suggestion
          if (err) {
           cb(err)
          } else if (getIssuerCertificate(socket).fingerprint256 !== 'FO:OB:AR') {
            socket.destroy()
           cb(new Error('Fingerprint does not match'))
          } else {
            cb(null, socket)
          }
```

nit

**Review by mcollina — COMMENTED** [review]
Is there a reason why we have a class in the first place? I would just remove it and expose a function.

**Comment by ronag** [comment]
> Is there a reason why we have a class in the first place? I would just remove it and expose a function.

Nothing other than convenience of implementation.

**Comment by delvedor** [comment]
Oks, this pr is good to go from my point of view, the types are now correct.

**Comment by delvedor** [comment]
@ronag @mcollina I've changed the connector from class to function, let me know if it's what you had in mind.

**Review by mcollina — CHANGES_REQUESTED** [review]
Awesome work! Could you add docs?

**Review by ronag — APPROVED** [review]

File: examples/ca-fingerprint/index.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    data.body.on('end', () => {
+      console.log(Buffer.concat(bufs).toString('utf8'))
+      client.close()
+      server.close()
+    })
+  })
+})
+
+function getIssuerCertificate (socket) {
+  let certificate = socket.getPeerCertificate(true)
```
Comment (ronag) [inline examples/ca-fingerprint/index.js:56]: Should probably check that socket is a `tls.Socket`.

File: index.js [hunk 3]
```diff
@@ -20,6 +21,7 @@ module.exports.Client = Client
 module.exports.Pool = Pool
 module.exports.Agent = Agent
 
+module.exports.buildConnector = buildConnector
```
Comment (ronag) [inline index.js:24]: Now I see why you preferred a class 😄 .

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by mcollina** [comment]
Docs! ;)

**Comment by delvedor** [comment]
Ready to go!

File: docs/api/Client.md [hunk 4]
```diff
@@ -48,6 +48,31 @@ import { Client } from 'undici'
 const client = new Client('http://localhost:3000')
 ```
 
+### Example - Custom connector
+
+This will allow you to perform some additional check on the socket that will be used for the next request.
+
+```js
+'use strict'
+import { Client, buildConnector } from 'undici'
```
Comment (mcollina) [inline docs/api/Client.md:57]: we should document `buildConnect()` somewhere else as well.

Comment (delvedor) [reply docs/api/Client.md:57]: Done!

**Review by mcollina — APPROVED** [review]
lgtm

**Pull request merged**